### DISCUSSION TITLE
Consider not suppressing the raised exception when loading a factory

### DISCUSSION THREAD

**Discussion by jessekv** [Kludex/uvicorn#2018] [Ideas]
Hi,

Recently came across this error:

![MicrosoftTeams-image](https://github.com/encode/uvicorn/assets/24944371/2897f5ff-fdeb-45d0-a921-b2792d65868d)

Unfortunately its not much to go on. For any non-trivial factory, capturing and suppressing all TypeError makes it hard to debug:

https://github.com/encode/uvicorn/blob/ffa5b1ac96b10976ed0e092a0bc1dd5526101356/uvicorn/config.py#L470-L475

Would it make sense to raise the exception? Or use `inspect` rather than try-except?

**Comment by Kludex** [comment]
How would the full stack traceback be helpful here?

**Reply by jessekv** [reply]
The factory is a real function, and it is getting imported and called correctly. Inside the factory, a module is imported. While parsing some configuration, the module raises a TypeError. It is ambiguous which part of the configuration is incorrect, but a stack trace would make it clear exactly what is going wrong.

**Reply by jessekv** [reply]
Note that if it had raised any other exception (e.g. KeyError), there would be a stack trace. So uvicorn is treating TypeError as a special case and it suppresses them, because it does not distinguish between TypeErrors raised by the factory and the TypeError raised when attempting to call a non-callable app.

**Reply by jessekv** [reply]
I'd be happy to submit a PR if that helps clarify.

**Reply by mjog** [reply]
> How would the full stack traceback be helpful here?

The factory is a function, the function throws an error when invoked, and your psychic debugger is not functioning at the moment so you don't know where in the function's code path it is failing.

**Comment by jessekv** [comment]
It seems to me there are two options:
1. Keep the try/except, but reraise the exception instead of calling `sys.exit(1)`
2. use `inspect` to determine if its worth calling `loaded_app`, and not capture any raised exceptions
